### PR TITLE
Introduce model filtering by server mode

### PR DESCRIPTION
**Pull request by RyanSkonnord**
Builds on #34981. An early sketch of a way to ensure that each "silo" accesses only the database tables that are meant to exist in that silo. It leaves the actual tables alone, but hacks the affected model classes to raise an exception if any application code attempts to access them.

During the development phase, this could be an easy way to hunt for unexpected dependencies in a dev environment before we've finished the actual shape of the "siloed" persistence tier. We may want to keep some version of it in a production solution, if only to get clearer errors if a table is accessed incorrectly.

Still left to do: a way to tag "read-only" availability, such that it fails only on destructive operations.

### CODE REVIEW

**Comment by markstory** [comment]
> Still left to do: a way to tag "read-only" availability, such that it fails only on destructive operations.

We're certainly going to need this as Customer Silo instances will need read-only access to data that will be replicated across regions like users.

File: src/sentry/db/models/base.py [hunk 1]
```diff
@@ -129,3 +131,29 @@ def __model_class_prepared(sender: Any, **kwargs: Any) -> None:
...
+    from . import BaseQuerySet
+
+    def decorator(decorated_class: Type[BaseModel]) -> Type[BaseModel]:
+        def queryset_override() -> BaseQuerySet:
+            raise ServerModeDataError(
+                f"{decorated_class.__name__} is available only in the {mode.value} server mode"
+            )
+
+        if not issubclass(decorated_class, BaseModel):
+            raise ValueError("`@available_on` must decorate a Model class")
```
Comment (markstory) [inline src/sentry/db/models/base.py:151]: Once we get this merged and applied to existing tables, we should consider having a consistency test to ensure that all models now and in the future have this annotation applied.

**Review by markstory — COMMENTED** [review]
Looking good. I like the annotation based approach :clap:

**Comment by RyanSkonnord** [comment]
Re https://github.com/getsentry/sentry/pull/35499/commits/693d9feaf7372d5e50898fefd52bf1dd0107c5cf:

My first pass had a subtle flaw. Monkey-patching the manager object works fine for models such as `Organization` that set a custom manager object such as `OrganizationManager`. I thought it would work for other models, on the mistaken assumption that each one would have an individual `BaseManager` object as the value of its `objects` class variable. This is not the case. Models that extend from `BaseModel` without overriding `objects` all inherit the same `objects` value, which means we're monkey-patching the instance that they all share. Whichever decorator is evaluated last will win.

Looking into tweaks to my solution. I've adjusted the unit test so that it exposes the problem in the meantime.

**Comment by RyanSkonnord** [comment]
Much better. 😌

https://github.com/getsentry/sentry/blob/18afcc9f16d71008da0224f422ef7807f922173a/src/sentry/db/models/base.py#L155-L161

File: src/sentry/db/models/base.py [hunk 2]
```diff
@@ -129,3 +131,35 @@ def __model_class_prepared(sender: Any, **kwargs: Any) -> None:
...
+        assert isinstance(decorated_class.objects, BaseManager)
+
+        if not mode.is_active():
+            manager_class = type(decorated_class.objects)
+            manager_class_with_override = type(
+                manager_class.__name__,
+                (manager_class,),
+                {"get_queryset": queryset_override},
+            )
```
Comment (markstory) [inline src/sentry/db/models/base.py:160]: Neat stuff :clap:

**Comment by RyanSkonnord** [comment]
Folded into https://github.com/getsentry/sentry/pull/35994

**Pull request closed without merging**